File: app/src/main/java/com/proxy/shadowsocksr/impl/UDPRelayServer.java
package com.proxy.shadowsocksr.impl;

import android.util.Log;
import android.util.LruCache;

import com.proxy.shadowsocksr.impl.interfaces.OnNeedProtectUDPListener;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.SocketAddress;
import java.nio.ByteBuffer;
import java.nio.channels.DatagramChannel;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public final class UDPRelayServer extends Thread
{
    private DatagramChannel udpServer;
    private InetSocketAddress isaLocal;
    private InetSocketAddress isaRemote;
    private UDPEncryptor crypto;

    private String remoteIP;
    private String localIP;
    private int remotePort;
    private int localPort;

    private final int ivLen;

    private ExecutorService exec;
    private volatile boolean isRunning = true;

    private LruCache<SocketAddress, UDPRemoteDataHandler> cache;

    private OnNeedProtectUDPListener onNeedProtectUDPListener;

    public UDPRelayServer(String remoteIP, String localIP, int remotePort, int localPort,
            String cryptMethod, String pwd)
    {
        this.remoteIP = remoteIP;
        this.localIP = localIP;
        this.remotePort = remotePort;
        this.localPort = localPort;
        cache = new LruCache<>(64);//save ram
        crypto = new UDPEncryptor(pwd, cryptMethod);
        ivLen = crypto.getIVLen();
    }

    public void setOnNeedProtectUDPListener(OnNeedProtectUDPListener onNeedProtectUDPListener)
    {
        this.onNeedProtectUDPListener = onNeedProtectUDPListener;
    }

    public void stopUDPRelayServer()
    {
        isRunning = false;
        //
        try
        {
            udpServer.socket().close();
            udpServer.close();
        }
        catch (Exception ignored)
        {
        }
        if (cache.size() > 0)
        {
            cache.evictAll();
        }
        udpServer = null;
        exec.shutdown();
    }

    @Override public void run()
    {
        isaLocal = new InetSocketAddress(localIP, localPort);
        isaRemote = new InetSocketAddress(remoteIP, remotePort);

        exec = Executors.newCachedThreadPool();
        //new ThreadPoolExecutor(1, Integer.MAX_VALUE, 300L, TimeUnit.SECONDS, new SynchronousQueue<Runnable>());

        ByteBuffer buf = ByteBuffer.allocate(1472);

        while (isRunning)//When udp server crashed, prepare it again.
        {
            try
            {
                udpServer = DatagramChannel.open();
                udpServer.configureBlocking(true);
                udpServer.socket().bind(isaLocal);
            }
            catch (Exception e)
            {
                Log.e("EXC", "UDPRealyServer Init Failed!");
                return;
            }
            //
            try
            {
                while (isRunning)
                {
                    SocketAddress localAddress = udpServer.receive(buf);
                    //
                    buf.flip();
                    int rcnt = buf.limit();
                    if (rcnt < 8)
                    {
                        Log.e("EXC", "LOCAL RECV SMALL PKG");
                        buf.clear();//not response small package
                        continue;
                    }
                    //
                    if (!(buf.get() == 0 && //RSV
                          buf.get() == 0 && //RSV
                          buf.get() == 0))  //FRAG
                    {
                        Log.e("EXC", "LOCAL RECV NOT SOCKS5 UDP PKG");
                        buf.clear();
                        continue;
                    }
                    //
                    byte[] dataLocalIn = new byte[rcnt - 3];
                    buf.get(dataLocalIn);
                    byte[] dataRemoteOut = crypto.encrypt(dataLocalIn);
                    //
                    UDPRemoteDataHandler handler = cache.get(localAddress);
                    if (handler == null)
                    {
                        DatagramChannel remoteChannel = DatagramChannel.open();
                        remoteChannel.configureBlocking(true);
                        remoteChannel.connect(isaRemote);
                        boolean isProtected = onNeedProtectUDPListener
                                .onNeedProtectUDP(remoteChannel.socket());
                        if (isProtected)
                        {
                            handler = new UDPRemoteDataHandler(localAddress, remoteChannel);
                            cache.put(localAddress, handler);
                            exec.submit(handler);
                        }
                        else
                        {
                            buf.clear();
                            continue;
                        }
                    }
                    handler.remoteChannel.write(ByteBuffer.wrap(dataRemoteOut));
                    //
                    buf.clear();
                }
            }
            catch (Exception e)
            {
                Log.e("EXC", "UDPRealyServer EXEC: " + e.getMessage());
            }
            try
            {
                udpServer.close();
            }
            catch (Exception ignored)
            {
            }
            buf.clear();
        }
    }

    class UDPRemoteDataHandler implements Runnable
    {
        public SocketAddress localAddress;
        public DatagramChannel remoteChannel;

        public ByteBuffer remoteReadBuf = ByteBuffer.allocate(1500);

        public UDPRemoteDataHandler(SocketAddress localAddress,
                DatagramChannel remoteChannel)
        {
            this.localAddress = localAddress;
            this.remoteChannel = remoteChannel;
        }

        @Override public void run()
        {
            try
            {
                while (isRunning)
                {
                    int rcnt = remoteChannel.read(remoteReadBuf);
                    if (rcnt < ivLen + 8)
                    {
                        remoteReadBuf.clear();//not response small package, just drop.
                        continue;
                    }
                    remoteReadBuf.flip();
                    byte[] dataRemoteIn = new byte[rcnt];
                    remoteReadBuf.get(dataRemoteIn);
                    dataRemoteIn = crypto.decrypt(dataRemoteIn);
                    byte[] dataLocalOut = new byte[dataRemoteIn.length + 3];
                    System.arraycopy(dataRemoteIn, 0, dataLocalOut, 3, dataRemoteIn.length);
                    udpServer.send(ByteBuffer.wrap(dataLocalOut), localAddress);
                    //
                    remoteReadBuf.clear();
                }
            }
            catch (Exception e)
            {
                Log.e("EXC", "UDP REMOTE EXC: " + e.getMessage());
                cache.remove(localAddress);
                try
                {
                    remoteChannel.close();
                }
                catch (IOException ignored)
                {
                }
                remoteChannel = null;
            }
        }
    }
}

File: app/src/main/java/com/proxy/shadowsocksr/impl/UDPRelayServer.kt
package com.proxy.shadowsocksr.impl

import android.util.Log
import android.util.LruCache
import com.proxy.shadowsocksr.impl.interfaces.OnNeedProtectUDPListener
import java.io.IOException
import java.net.*
import java.nio.ByteBuffer
import java.nio.channels.DatagramChannel
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

class UDPRelayServer(
        private val remoteIP: String, private val localIP: String, private val remotePort: Int,
        private val localPort: Int, private val isTunnelMode: Boolean, private val isVPNMode: Boolean,
        cryptMethod: String, pwd: String, dnsIp: String?, dnsPort: Int?) : Thread()
{
    private var udpServer: DatagramSocket? = null
    private var isaLocal: InetSocketAddress? = null
    private var isaRemote: InetSocketAddress? = null
    private val crypto: UDPEncryptor

    private val ivLen: Int

    private var exec: ExecutorService? = null
    @Volatile private var isRunning = true

    private val cache: LruCache<SocketAddress, UDPRemoteDataHandler>

    private val targetDnsHead = ByteArray(7)

    var onNeedProtectUDPListener: OnNeedProtectUDPListener? = null

    init
    {
        cache = LruCache<SocketAddress, UDPRemoteDataHandler>(48)
        crypto = UDPEncryptor(pwd, cryptMethod)
        ivLen = crypto.ivLen
        if (isTunnelMode)
        {
            //UnknownHostException? don't be silly
            var dnsIP = InetAddress.getByName(dnsIp).address
            //
            targetDnsHead[0] = 1
            System.arraycopy(dnsIP, 0, targetDnsHead, 1, 4)
            targetDnsHead[5] = ((dnsPort!!.shr(8)) and 0xFF).toByte()
            targetDnsHead[6] = (dnsPort and 0xFF).toByte()
        }
    }

    fun stopUDPRelayServer()
    {
        isRunning = false
        //
        try
        {
            udpServer!!.close()
        }
        catch (ignored: Exception)
        {
        }

        cache.evictAll()
        udpServer = null
        exec!!.shutdown()
    }

    override fun run()
    {
        isaLocal = InetSocketAddress(localIP, localPort)
        isaRemote = InetSocketAddress(remoteIP, remotePort)

        exec = Executors.newCachedThreadPool()
        //new ThreadPoolExecutor(1, Integer.MAX_VALUE, 300L, TimeUnit.SECONDS, new SynchronousQueue<Runnable>());

        val buf = ByteArray(1472)
        var buff: DatagramPacket?

        while (isRunning)
        {
            //When udp server crashed, prepare it again.
            try
            {
                udpServer = DatagramSocket(isaLocal)
            }
            catch (e: Exception)
            {
                Log.e("EXC", "UDPRelayServer Init Failed!")
                return
            }
            //
            try
            {
                while (isRunning)
                {
                    buff = DatagramPacket(buf, 0, buf.size)
                    udpServer!!.receive(buff)
                    //
                    val dataLocalIn: ByteArray
                    //
                    if (isTunnelMode)
                    {
                        //direct build target dns server socks5 request pkg
                        if (buff.length < 12)
                        {
                            Log.e("EXC", "LOCAL RECV SMALL PKG")
                            continue
                        }
                        dataLocalIn = ByteArray(7 + buff.length)
                        System.arraycopy(targetDnsHead, 0, dataLocalIn, 0, 7)
                        System.arraycopy(buf, 0, dataLocalIn, 7, buff.length)
                    }
                    else
                    {
                        if (buff.length < 8)
                        {
                            Log.e("EXC", "LOCAL RECV SMALL PKG")
                            continue
                        }
                        //
                        if (!(buf[0] == 0.toByte() && //RSV
                                buf[1] == 0.toByte() && //RSV
                                buf[2] == 0.toByte()))  //FRAG
                        {
                            Log.e("EXC", "LOCAL RECV NOT SOCKS5 UDP PKG")
                            continue
                        }
                        //
                        dataLocalIn = ByteArray(buff.length - 3)
                        System.arraycopy(buf, 3, dataLocalIn, 0, buff.length - 3)
                    }
                    val dataRemoteOut = crypto.encrypt(dataLocalIn)
                    //
                    var handler: UDPRemoteDataHandler? = cache.get(buff.socketAddress)
                    if (handler == null)
                    {
                        val remoteSkt = DatagramSocket()
                        remoteSkt.connect(isaRemote)
                        if (isVPNMode)
                        {
                            val isProtected = onNeedProtectUDPListener!!.onNeedProtectUDP(remoteSkt)
                            if (isProtected)
                            {
                                handler = UDPRemoteDataHandler(buff.socketAddress, remoteSkt)
                                cache.put(buff.socketAddress, handler)
                                exec!!.execute(handler)
                            }
                            else
                            {
                                continue
                            }
                        }
                        else
                        {
                            //Nat mode need not protect
                            handler = UDPRemoteDataHandler(buff.socketAddress, remoteSkt)
                            cache.put(buff.socketAddress, handler)
                            exec!!.execute(handler)
                        }
                    }

                    val pktRemoteOut = DatagramPacket(dataRemoteOut, 0, dataRemoteOut.size)
                    //pktRemoteOut.socketAddress = isaRemote
                    handler.remoteSkt!!.send(pktRemoteOut)
                }
            }
            catch (e: Exception)
            {
                Log.e("EXC", "UDPRealyServer EXEC: " + e.message)
            }

            try
            {
                udpServer!!.close()
            }
            catch (ignored: Exception)
            {
            }
        }
    }

    inner class UDPRemoteDataHandler(val localAddress: SocketAddress,
            var remoteSkt: DatagramSocket?) : Runnable
    {

        var buf = ByteArray(1472)
        var buff: DatagramPacket? = null

        override fun run()
        {
            try
            {
                while (isRunning)
                {
                    buff = DatagramPacket(buf, 0, buf.size)
                    remoteSkt!!.receive(buff)
                    //
                    if (buff!!.length < ivLen + 1)
                    {
                        //continue
                        break
                    }

                    var dataRemoteIn = ByteArray(buff!!.length)
                    System.arraycopy(buf, 0, dataRemoteIn, 0, buff!!.length)
                    dataRemoteIn = crypto.decrypt(dataRemoteIn)
                    //
                    val dataLocalOut: ByteArray
                    if (isTunnelMode)
                    {
                        dataLocalOut = dataRemoteIn
                    }
                    else
                    {
                        dataLocalOut = ByteArray(dataRemoteIn.size + 3)
                        System.arraycopy(dataRemoteIn, 0, dataLocalOut, 3, dataRemoteIn.size)
                    }
                    //
                    val pktLocalOut = DatagramPacket(dataLocalOut, 0, dataLocalOut.size)
                    pktLocalOut.socketAddress = localAddress
                    udpServer!!.send(pktLocalOut)
                }
            }
            catch (e: Exception)
            {
                Log.e("EXC", "UDP REMOTE EXC: " + e.message)
                try
                {
                    cache.remove(localAddress)
                    remoteSkt!!.close()
                }
                catch (ignored: IOException)
                {
                }
                remoteSkt = null
            }
        }
    }
}